File: experiments/adrianco/experiment-10/brazil/runs/language=rust_model=claude-fable-5/rep1/tests/sample_questions.rs
//! The specification's success criteria require that at least 20 sample
//! questions can be answered. Each test here phrases one natural-language
//! question as the MCP tool call an LLM would issue, runs it through the
//! real tool dispatch (`mcp::call_tool`), and checks the answer text.

use std::path::Path;
use std::sync::OnceLock;

use brazilian_soccer_mcp::data::Store;
use brazilian_soccer_mcp::mcp::call_tool;
use serde_json::json;

fn store() -> &'static Store {
    static STORE: OnceLock<Store> = OnceLock::new();
    STORE.get_or_init(|| {
        Store::load(Path::new(env!("CARGO_MANIFEST_DIR")).join("data/kaggle").as_path())
            .expect("datasets must load")
    })
}

fn ask(tool: &str, args: serde_json::Value) -> String {
    call_tool(store(), tool, &args).expect("tool call must succeed")
}

// 1. "Show me all Flamengo vs Fluminense matches"
#[test]
fn q01_flamengo_vs_fluminense() {
    let a = ask(
        "search_matches",
        json!({"team": "Flamengo", "opponent": "Fluminense", "limit": 50}),
    );
    assert!(a.contains("Flamengo") && a.contains("Fluminense"));
    assert!(!a.contains("No matches found"));
}

// 2. "What matches did Palmeiras play in 2023?"
#[test]
fn q02_palmeiras_2023() {
    let a = ask("search_matches", json!({"team": "Palmeiras", "season": 2023}));
    assert!(a.contains("Palmeiras"));
    assert!(a.contains("2023-"));
}

// 3. "Find all Copa do Brasil finals"
#[test]
fn q03_copa_do_brasil_finals() {
    let a = ask(
        "search_matches",
        json!({"competition": "Copa do Brasil", "stage": "final", "limit": 40}),
    );
    assert!(a.contains("Copa do Brasil"));
    assert!(a.contains("Final"));
}

// 4. "What is Corinthians' home record in 2022?"
#[test]
fn q04_corinthians_home_2022() {
    let a = ask(
        "team_stats",
        json!({"team": "Corinthians", "season": 2022, "competition": "Serie A", "venue": "home"}),
    );
    assert!(a.contains("Corinthians"));
    assert!(a.contains("Matches: 19"));
    assert!(a.contains("Win rate"));
}

// 5. "Which team scored the most goals in Serie A 2023?"
#[test]
fn q05_top_scoring_team_2023() {
    let a = ask("competition_stats", json!({"competition": "Serie A", "season": 2023}));
    assert!(a.contains("Top scoring teams"));
    assert!(a.contains("Average goals per match"));
}

// 6. "Compare Palmeiras and Santos head-to-head"
#[test]
fn q06_palmeiras_santos_h2h() {
    let a = ask("head_to_head", json!({"team1": "Palmeiras", "team2": "Santos"}));
    assert!(a.contains("Palmeiras") && a.contains("Santos"));
    assert!(a.contains("wins") && a.contains("draws"));
}

// 7. "Find all Brazilian players in the dataset"
#[test]
fn q07_brazilian_players() {
    let a = ask("search_players", json!({"nationality": "Brazil"}));
    assert!(a.contains("Found 827 players"));
    assert!(a.contains("Neymar Jr"));
}

// 8. "Who are the highest-rated players at Grêmio?" (accent-insensitive)
#[test]
fn q08_top_players_at_gremio() {
    let a = ask("search_players", json!({"club": "Gremio", "limit": 5}));
    assert!(a.contains("Grêmio"));
    assert!(a.contains("Overall"));
}

// 9. "Show me all forwards from Santos"
#[test]
fn q09_santos_forwards() {
    let a = ask("search_players", json!({"club": "Santos", "position": "forward"}));
    assert!(a.contains("Santos"));
    assert!(!a.contains("No players found"));
}

// 10. "Who won the 2019 Brasileirão?"
#[test]
fn q10_2019_champion() {
    let a = ask("league_standings", json!({"season": 2019}));
    assert!(a.contains(" 1. Flamengo - 90 pts"));
    assert!(a.contains("Champion"));
}

// 11. "Show the 2018 Copa Libertadores knockout results"
#[test]
fn q11_2018_libertadores() {
    let a = ask(
        "search_matches",
        json!({"competition": "Libertadores", "season": 2018, "stage": "final"}),
    );
    assert!(a.contains("Copa Libertadores"));
    assert!(!a.contains("No matches found"));
}

// 12. "Which teams were relegated in 2020?"
#[test]
fn q12_relegated_2020() {
    let a = ask("league_standings", json!({"season": 2020}));
    assert!(a.contains("Relegated"));
    // The real bottom four of 2020: Vasco, Goiás, Coritiba, Botafogo
    assert!(a.contains("Vasco da Gama") && a.contains("Botafogo"));
}

// 13. "What's the average goals per match in the Brasileirão?"
#[test]
fn q13_average_goals() {
    let a = ask("competition_stats", json!({"competition": "Brasileirão"}));
    let line = a
        .lines()
        .find(|l| l.contains("Average goals per match"))
        .expect("answer must include the average");
    let value: f64 = line.rsplit(' ').next().unwrap().parse().unwrap();
    assert!((1.5..=3.5).contains(&value), "implausible average {value}");
}

// 14. "Which team has the best away record?" (via venue-filtered stats)
#[test]
fn q14_away_record() {
    let a = ask(
        "team_stats",
        json!({"team": "Flamengo", "season": 2019, "competition": "Serie A", "venue": "away"}),
    );
    assert!(a.contains("away record"));
    assert!(a.contains("Matches: 19"));
}

// 15. "Show me the biggest wins in the dataset"
#[test]
fn q15_biggest_wins() {
    let a = ask("biggest_wins", json!({"limit": 5}));
    assert!(a.contains("Biggest victories"));
    // The top margin must be at least six goals somewhere in 18k matches.
    assert!(!a.contains("No matches found"));
}

// 16. "When did Flamengo last play Corinthians?"
#[test]
fn q16_last_fla_corinthians() {
    let a = ask(
        "search_matches",
        json!({"team": "Flamengo", "opponent": "Corinthians", "limit": 1}),
    );
    // Most recent first; the answer carries date and score.
    assert!(a.contains("Flamengo") && a.contains("Corinthians"));
    assert!(a.lines().any(|l| l.starts_with("- 20")));
}

// 17. "Who is Gabriel Barbosa?" (player lookup by partial name)
#[test]
fn q17_who_is_player() {
    let a = ask("player_info", json!({"name": "Casemiro"}));
    assert!(a.contains("Casemiro"));
    assert!(a.contains("Overall"));
    assert!(a.contains("Nationality: Brazil"));
}

// 18. "Which players play for Sport Club do Recife?"
#[test]
fn q18_players_by_club() {
    let a = ask("search_players", json!({"club": "Sport Club do Recife"}));
    assert!(!a.contains("No players found"));
}

// 19. "Show me all derbies in 2023" (Fla-Flu as the classic example)
#[test]
fn q19_derbies_2023() {
    let a = ask(
        "search_matches",
        json!({"team": "Flamengo", "opponent": "Fluminense", "season": 2023}),
    );
    assert!(!a.contains("No matches found"));
    assert!(a.contains("2023-"));
}

// 20. "What competitions has Palmeiras played in?"
#[test]
fn q20_palmeiras_competitions() {
    let a = ask("team_stats", json!({"team": "Palmeiras"}));
    assert!(a.contains("By competition:"));
    assert!(a.contains("Brasileirão Série A"));
    assert!(a.contains("Copa Libertadores"));
    assert!(a.contains("Copa do Brasil"));
}

// 21. "Which team has the best home record?" (league-wide aggregate)
#[test]
fn q21_home_advantage() {
    let a = ask("competition_stats", json!({"competition": "Serie A"}));
    assert!(a.contains("Home wins"));
    assert!(a.contains("Away wins"));
}

// 22. "Who are the top Brazilian players?"
#[test]
fn q22_top_brazilians() {
    let a = ask(
        "search_players",
        json!({"nationality": "Brazil", "min_overall": 85, "limit": 10}),
    );
    assert!(a.contains("Neymar Jr"));
    assert!(!a.contains("No players found"));
}

// 23. "Compare the 2018 and 2019 seasons"
#[test]
fn q23_compare_seasons() {
    let a18 = ask("competition_stats", json!({"competition": "Serie A", "season": 2018}));
    let a19 = ask("competition_stats", json!({"competition": "Serie A", "season": 2019}));
    assert!(a18.contains("Average goals per match"));
    assert!(a19.contains("Average goals per match"));
    assert_ne!(a18, a19);
}

// 24. "What was the score?" (follow-up: a specific fixture has its score)
#[test]
fn q24_specific_score() {
    let a = ask(
        "search_matches",
        json!({"team": "Flamengo", "opponent": "Santos", "season": 2019, "limit": 2}),
    );
    // Each line shows "Team X-Y Team".
    assert!(a.lines().any(|l| l.contains("Flamengo") && l.contains("Santos")));
}

// 25. "What data do you have?" (coverage / capability question)
#[test]
fn q25_data_coverage() {
    let a = ask("list_competitions", json!({}));
    assert!(a.contains("Brasileirão Série A"));
    assert!(a.contains("Copa do Brasil"));
    assert!(a.contains("Copa Libertadores"));
    assert!(a.contains("fifa_data.csv"));
}

File: experiments/adrianco/experiment-46-opus5/brazil/runs/language=elixir_model=claude-opus-5_prompt=neutral/rep1/lib/brazilian_soccer/sample_questions.ex
defmodule BrazilianSoccer.SampleQuestions do
  @moduledoc """
  The catalogue of questions this server can answer, each mapped to the MCP
  tool call that answers it.

  It is used three ways: as the `brazilian-soccer://sample-questions`
  resource (so a connected model can see what is possible), as the script for
  `mix soccer.demo`, and as the acceptance test in
  `test/sample_questions_test.exs`.
  """

  alias BrazilianSoccer.MCP.Tools

  @questions [
    # --- 1. match queries --------------------------------------------------
    %{
      category: "Match queries",
      question: "Show me all Flamengo vs Fluminense matches",
      tool: "head_to_head",
      arguments: %{"team_a" => "Flamengo", "team_b" => "Fluminense"}
    },
    %{
      category: "Match queries",
      question: "What matches did Palmeiras play in 2023?",
      tool: "search_matches",
      arguments: %{"team" => "Palmeiras", "season" => 2023, "limit" => 10}
    },
    %{
      category: "Match queries",
      question: "Find all Copa do Brasil finals",
      tool: "search_matches",
      arguments: %{"competition" => "Copa do Brasil", "stage" => "final", "limit" => 30}
    },
    %{
      category: "Match queries",
      question: "When did Flamengo last play Corinthians, and what was the score?",
      tool: "last_meeting",
      arguments: %{"team_a" => "Flamengo", "team_b" => "Corinthians"}
    },
    %{
      category: "Match queries",
      question: "Show me all derbies in 2023",
      tool: "find_derbies",
      arguments: %{"season" => 2023}
    },
    %{
      category: "Match queries",
      question: "Which matches did Grêmio play away in the 2019 Libertadores?",
      tool: "search_matches",
      arguments: %{
        "team" => "Grêmio",
        "competition" => "Libertadores",
        "season" => 2019,
        "venue" => "away"
      }
    },

    # --- 2. team queries ---------------------------------------------------
    %{
      category: "Team queries",
      question: "What is Corinthians' home record in 2022?",
      tool: "team_stats",
      arguments: %{
        "team" => "Corinthians",
        "season" => 2022,
        "competition" => "Brasileirão",
        "venue" => "home"
      }
    },
    %{
      category: "Team queries",
      question: "Which team scored the most goals in Série A 2023?",
      tool: "competition_summary",
      arguments: %{"competition" => "Serie A", "season" => 2023}
    },
    %{
      category: "Team queries",
      question: "Compare Palmeiras and Santos head-to-head",
      tool: "compare_teams",
      arguments: %{"team_a" => "Palmeiras", "team_b" => "Santos"}
    },
    %{
      category: "Team queries",
      question: "Which team has the best home record in the Brasileirão?",
      tool: "team_rankings",
      arguments: %{"metric" => "home", "competition" => "Brasileirão", "limit" => 10}
    },
    %{
      category: "Team queries",
      question: "Which team has the best away record?",
      tool: "team_rankings",
      arguments: %{"metric" => "away", "competition" => "Brasileirão", "limit" => 10}
    },
    %{
      category: "Team queries",
      question: "What competitions has Palmeiras played in?",
      tool: "team_profile",
      arguments: %{"team" => "Palmeiras"}
    },
    %{
      category: "Team queries",
      question: "Is \"Atlético\" Mineiro, Goianiense or Paranaense?",
      tool: "resolve_team_name",
      arguments: %{"name" => "Atlético"}
    },

    # --- 3. player queries -------------------------------------------------
    %{
      category: "Player queries",
      question: "Find all Brazilian players in the dataset",
      tool: "players_by_nationality",
      arguments: %{"nationality" => "Brazil", "limit" => 10}
    },
    %{
      category: "Player queries",
      question: "Who are the highest-rated players at Grêmio?",
      tool: "club_squad",
      arguments: %{"club" => "Grêmio"}
    },
    %{
      category: "Player queries",
      question: "Show me all forwards from Santos",
      tool: "search_players",
      arguments: %{"club" => "Santos", "position_group" => "forward"}
    },
    %{
      category: "Player queries",
      question: "Who is Neymar?",
      tool: "player_profile",
      arguments: %{"name" => "Neymar"}
    },
    %{
      category: "Player queries",
      question: "Which players play for Flamengo?",
      tool: "club_squad",
      arguments: %{"club" => "Flamengo"}
    },
    %{
      category: "Player queries",
      question: "Who are the top rated Brazilian players?",
      tool: "search_players",
      arguments: %{"nationality" => "Brazil", "sort_by" => "overall", "limit" => 10}
    },
    %{
      category: "Player queries",
      question: "Which Brazilian goalkeepers are rated 80 or higher?",
      tool: "search_players",
      arguments: %{
        "nationality" => "Brazil",
        "position_group" => "goalkeeper",
        "min_overall" => 80
      }
    },

    # --- 4. competition queries -------------------------------------------
    %{
      category: "Competition queries",
      question: "Who won the 2019 Brasileirão?",
      tool: "competition_champion",
      arguments: %{"competition" => "Brasileirão", "season" => 2019}
    },
    %{
      category: "Competition queries",
      question: "Show the 2019 Brasileirão final standings",
      tool: "league_standings",
      arguments: %{"competition" => "Brasileirão", "season" => 2019}
    },
    %{
      category: "Competition queries",
      question: "Which teams were relegated in 2020?",
      tool: "league_standings",
      arguments: %{"competition" => "Brasileirão", "season" => 2020}
    },
    %{
      category: "Competition queries",
      question: "Show the 2018 Copa Libertadores bracket",
      tool: "cup_bracket",
      arguments: %{"competition" => "Libertadores", "season" => 2018}
    },
    %{
      category: "Competition queries",
      question: "Who won the 2019 Copa do Brasil?",
      tool: "competition_champion",
      arguments: %{"competition" => "Copa do Brasil", "season" => 2019}
    },
    %{
      category: "Competition queries",
      question: "Which competitions and seasons are covered?",
      tool: "list_competitions",
      arguments: %{}
    },

    # --- 5. statistical analysis ------------------------------------------
    %{
      category: "Statistical analysis",
      question: "What's the average goals per match in the Brasileirão?",
      tool: "match_statistics",
      arguments: %{"competition" => "Brasileirão"}
    },
    %{
      category: "Statistical analysis",
      question: "Show me the biggest wins in the dataset",
      tool: "biggest_wins",
      arguments: %{"limit" => 10}
    },
    %{
      category: "Statistical analysis",
      question: "What were the highest scoring matches in the Brasileirão?",
      tool: "highest_scoring_matches",
      arguments: %{"competition" => "Brasileirão", "limit" => 5}
    },
    %{
      category: "Statistical analysis",
      question: "Compare the 2018 and 2019 seasons",
      tool: "compare_seasons",
      arguments: %{"competition" => "Brasileirão", "seasons" => [2018, 2019]}
    },
    %{
      category: "Statistical analysis",
      question: "How big is home advantage in the Brasileirão?",
      tool: "home_advantage",
      arguments: %{"competition" => "Brasileirão"}
    },
    %{
      category: "Statistical analysis",
      question: "What data is loaded into the knowledge graph?",
      tool: "list_datasets",
      arguments: %{}
    }
  ]

  @doc "Every sample question with the tool call that answers it."
  @spec all() :: [map]
  def all, do: @questions

  @doc "Sample questions grouped by category."
  @spec by_category() :: [{binary, [map]}]
  def by_category do
    @questions
    |> Enum.group_by(& &1.category)
    |> Enum.sort_by(fn {category, _} -> category end)
  end

  @doc "Answer one sample question through the MCP tool layer."
  @spec answer(map) :: {:ok, binary} | {:error, binary}
  def answer(%{tool: tool, arguments: arguments}) do
    case Tools.call(tool, arguments) do
      {:ok, %{text: text}} -> {:ok, text}
      {:error, message} -> {:error, message}
    end
  end

  @doc "Markdown listing, used as the `sample-questions` MCP resource."
  @spec markdown() :: binary
  def markdown do
    body =
      Enum.map_join(by_category(), "\n\n", fn {category, questions} ->
        lines =
          Enum.map_join(questions, "\n", fn question ->
            "- #{question.question} → `#{question.tool}` #{inspect(question.arguments)}"
          end)

        "## #{category}\n#{lines}"
      end)

    "# Questions this server can answer\n\n#{body}\n"
  end
end
